dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -24,62 +24,92 @@ namespace DB
 template <typename NestedTaskQueueType>
 void ResourceControlQueue<NestedTaskQueueType>::submit(TaskPtr && task)
 {
+    try
     {
-        std::lock_guard lock(mu);
-        submitWithoutLock(std::move(task));
+        {
+            std::lock_guard lock(mu);
+            // May throw resource group not found exception.
+            NestedTaskQueuePtr task_queue = getSubmitTaskQueuWithoutLock(task);
+            if likely (task_queue)
+                task_queue->submit(std::move(task));
+        }
+        cv.notify_one();
+    }
+    catch (...)
+    {
+        assert(task);
+        task->onErrorOccurred(std::current_exception());
     }
-    cv.notify_one();
 }
 
 template <typename NestedTaskQueueType>
 void ResourceControlQueue<NestedTaskQueueType>::submit(std::vector<TaskPtr> & tasks)
 {
+    if (tasks.empty())
+        return;
+
+    std::vector<std::pair<TaskPtr, std::exception_ptr>> error_task_infos;
     {
-        std::lock_guard lock(mu);
+        std::unique_lock lock(mu);
         for (auto & task : tasks)
         {
-            submitWithoutLock(std::move(task));
-            cv.notify_one();
+            try
+            {
+                // May throw resource group not found exception.
+                NestedTaskQueuePtr task_queue = getSubmitTaskQueuWithoutLock(task);
+                if likely (task_queue)
+                    task_queue->submit(std::move(task));
```

Comment:
You can use `// NOLINT(bugprone-use-after-move)`, so you don't need to modify the code

Reply:
```suggestion
                submitWithoutLock(std::move(task));  // NOLINT(bugprone-use-after-move)
```

Reply:
I think it's dangerous to do this?
After std::move, the state of Task is unspecified.

Reply:
Before `other = moved_obj` is called, the task remains unchanged, so it is safe